agent/engine/engine_unix_integ_test.go

```diff
@@ -1621,81 +1766,86 @@ func TestExecCommandAgent(t *testing.T) {
-
-	stateChangeEvents := taskEngine.StateChangeEvents()
-
-	taskArn := "testDockerExec"
-	testTask := createTestTask(taskArn)
-
-	A := createTestContainerWithImageAndName(baseImageForOS, "A")
-
-	A.EntryPoint = &entryPointForOS
-	A.Command = []string{"sleep 30"}
-	A.Essential = true
+func verifyMockExecCommandAgentIsRunning(t *testing.T, client *sdkClient.Client, containerId string) string {
+	return verifyMockExecCommandAgentStatus(t, client, containerId, "", true)
+}
 
-	testTask.Containers = []*apicontainer.Container{
-		A,
-	}
-	execConfig := types.ExecConfig{
-		Detach: true,
-		Cmd:    []string{"ls"},
-	}
-	go taskEngine.AddTask(testTask)
+func verifyMockExecCommandAgentIsStopped(t *testing.T, client *sdkClient.Client, containerId, prevPid string) {
+	verifyMockExecCommandAgentStatus(t, client, containerId, prevPid, false)
+}
 
-	ctx, cancel := context.WithCancel(context.TODO())
+func verifyMockExecCommandAgentStatus(t *testing.T, client *sdkClient.Client, containerId, prevPid string, checkIsRunning bool) string {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
 	defer cancel()
-
-	finished := make(chan interface{})
+	res := make(chan string, 1)
 	go func() {
-		// Both containers should start
-		verifyContainerRunningStateChange(t, taskEngine)
-		verifyTaskIsRunning(stateChangeEvents, testTask)
+		for {
+			top, err := client.ContainerTop(ctx, containerId, nil)
+			if err != nil {
+				continue
+			}
+			cmdPos := -1
+			pidPos := -1
+			for i, t := range top.Titles {
+				if strings.ToUpper(t) == "CMD" {
+					cmdPos = i
+				}
+				if strings.ToUpper(t) == "PID" {
+					pidPos = i
+				}
 
-		containerMap, _ := taskEngine.(*DockerTaskEngine).state.ContainerMapByArn(testTask.Arn)
-		dockerID := containerMap[testTask.Containers[0].Name].DockerID
+			}
+			require.NotEqual(t, -1, cmdPos, "CMD title not found in the container top response")
+			require.NotEqual(t, -1, pidPos, "PID title not found in the container top response")
+			for _, proc := range top.Processes {
+				if proc[cmdPos] == filepath.Join(execcmd.ContainerBinDir, execcmd.BinName) {
+					res <- proc[pidPos]
+					return
+				}
+			}
 
-		//Create Exec object on the host
-		execContainerOut, err := taskEngine.(*DockerTaskEngine).client.CreateContainerExec(ctx, dockerID, execConfig, dockerclient.ContainerExecCreateTimeout)
-		require.NoError(t, err)
-		require.NotNil(t, execContainerOut)
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(retry.AddJitter(time.Second, time.Second*5)):
+			}
+		}
+	}()
 
-		//Start the above Exec process on the host
-		err1 := taskEngine.(*DockerTaskEngine).client.StartContainerExec(ctx, execContainerOut.ID, dockerclient.ContainerExecStartTimeout)
-		require.NoError(t, err1)
+	var (
+		isRunning bool
+		pid       string
+	)
+	select {
+	case <-ctx.Done():
+	case r := <-res:
+		if r != "" {
+			pid = r
+			isRunning = true
+			if prevPid != "" && pid != prevPid {
+				isRunning = false
+			}
```

Comment:
nit (extremely nitty):
I would call `prevPid` something like `expectedPid` to avoid confusion.  I got tangled up a bit here trying to decide whether I actually want my pid to equal my prevPid.

Reply:
sure, will change before merging

Reply:
changed